src/solvers/twostep_time_solver.C

```diff
@@ -300,27 +301,67 @@ std::pair<unsigned int, Real> TwostepTimeSolver::adjoint_solve (const QoISet & q
   // Take the first adjoint 'half timestep'
   core_time_solver->adjoint_solve(qoi_indices);
 
+  // Record the sub step deltat we used for the last adjoint solve.
+  last_deltat = _system.deltat;
+
   // Adjoint advance the timestep
   core_time_solver->adjoint_advance_timestep();
 
   // The second half timestep
   std::pair<unsigned int, Real> full_adjoint_output = core_time_solver->adjoint_solve(qoi_indices);
 
-  // Tell our timesolver the combined deltat before adjoint advance timestep updates it
-  this->completedtimestep_deltat = 2.0*_system.deltat;
+  // Record the sub step deltat we used for the last adjoint solve.
+  last_deltat = _system.deltat;
 
-  // Adjoint advance again so solution history saves the adjoint
-  core_time_solver->adjoint_advance_timestep();
+  // Record the total size of the last timestep, for a 2StepTS, this is
+  // simply twice the deltat for each sub(half) step.
+  this->completed_timestep_size = 2.0*_system.deltat;
```

Comment:
We are totally handling system.deltat in an inconsistent way here, but that's an old design flaw and my fault, not anything wrong with this PR.